cpp/src/io/utilities/data_casting.cu

```diff
@@ -0,0 +1,952 @@
+          d_offsets[istring] = 0;
+        }
+        if constexpr (!is_warp) { __syncthreads(); }
+        break;  // gride-stride return;
+      }
+
+      // Skipping non-copied escaped characters
+      bool skip = !is_within_bounds;  // false;
+      // skip \ for \" \\ \/ \b \f \n \r \t \uXXXX
+      skip |= is_escaping_backslash;
+      if (is_within_bounds) {
+        // skip X for each X in \uXXXX
+        skip |= char_index - 2 >= 0 && is_slash.get_bit(warp_id, lane - 2) &&
+                in_begin[char_index - 1] == 'u';
+        skip |= char_index - 3 >= 0 && is_slash.get_bit(warp_id, lane - 3) &&
+                in_begin[char_index - 2] == 'u';
+        skip |= char_index - 4 >= 0 && is_slash.get_bit(warp_id, lane - 4) &&
+                in_begin[char_index - 3] == 'u';
+        skip |= char_index - 5 >= 0 && is_slash.get_bit(warp_id, lane - 5) &&
+                in_begin[char_index - 4] == 'u';
+      }
+      int this_num_out = 0;
+      cudf::char_utf8 write_char{};
+
+      if (!skip) {
+        // 1. Unescaped character
+        if (!is_prev_escaping_backslash) {
+          this_num_out = 1;
+          // writes char directly for non-unicode
+        } else {
+          // 2. Escaped character
+          if (escaped_char != UNICODE_SEQ) {
+            this_num_out = 1;
+            // writes char directly for non-unicode
+          } else {
+            // 3. Unicode
+            // UTF8 \uXXXX
+            auto hex_val     = parse_unicode_hex(in_begin + char_index + 1);
+            auto hex_low_val = 0;
+            // UTF16 \uXXXX\uXXXX
+            // Note: no need for scanned_backslash below because we already know that
+            // only '\u' check is enough.
+            if ((in_begin + char_index + UNICODE_HEX_DIGIT_COUNT + NUM_UNICODE_ESC_SEQ_CHARS) <
+                  in_end &&
+                in_begin[char_index + NUM_UNICODE_ESC_SEQ_CHARS - 1] == '\\' &&
+                in_begin[char_index + NUM_UNICODE_ESC_SEQ_CHARS] == 'u') {
+              hex_low_val = parse_unicode_hex(in_begin + char_index + 1 + 6);
+            }
+            if (hex_val >= UTF16_HIGH_SURROGATE_BEGIN && hex_val < UTF16_HIGH_SURROGATE_END &&
+                hex_low_val >= UTF16_LOW_SURROGATE_BEGIN && hex_low_val < UTF16_LOW_SURROGATE_END) {
+              // Compute UTF16-encoded code point
+              uint32_t unicode_code_point = 0x10000 +
+                                            ((hex_val - UTF16_HIGH_SURROGATE_BEGIN) << 10) +
+                                            (hex_low_val - UTF16_LOW_SURROGATE_BEGIN);
+              write_char   = strings::detail::codepoint_to_utf8(unicode_code_point);
+              this_num_out = strings::detail::bytes_in_char_utf8(write_char);
+            } else {
+              if (hex_val >= UTF16_LOW_SURROGATE_BEGIN && hex_val < UTF16_LOW_SURROGATE_END) {
+                skip         = true;
+                this_num_out = 0;
+                write_char   = 0;
+              } else {
```

Comment:
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question.  -->
**issue/question:** ‏This logic looks a _little_ bit different to the UTF-16 codepoint handling in `process_string` above. Is that deliberate? If not, which one is correct?

Specifically, in `process_string` there's no handling of `hex_val` being a "low" surrogate in the `else` case.

Reply:
OK, thanks